samza-core/src/main/java/org/apache/samza/zk/ZkMetadataStore.java

```diff
@@ -0,0 +1,117 @@
+
+import java.io.UnsupportedEncodingException;
+import java.util.concurrent.TimeUnit;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.I0Itec.zkclient.serialize.BytesPushThroughSerializer;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.ZkConfig;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.metadatastore.MetadataStore;
+import org.apache.samza.SamzaException;
+import org.I0Itec.zkclient.ZkClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of the {@link MetadataStore} interface where the
+ * metadata of the Samza job is stored in zookeeper.
+ */
+public class ZkMetadataStore implements MetadataStore {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ZkMetadataStore.class);
+
+  private final ZkClient zkClient;
+  private final ZkConfig zkConfig;
+  private final String zkBaseDir;
+
+  public ZkMetadataStore(Config config, MetricsRegistry metricsRegistry, String zkBaseDir) {
+    this.zkConfig = new ZkConfig(config);
+    this.zkClient = new ZkClient(zkConfig.getZkConnect(), zkConfig.getZkSessionTimeoutMs(), zkConfig.getZkConnectionTimeoutMs(), new BytesPushThroughSerializer());
+    this.zkBaseDir = zkBaseDir;
+    createIfNotExists(zkBaseDir);
+  }
+
+  @Override
+  public void init() {
+    zkClient.waitUntilConnected(zkConfig.getZkConnectionTimeoutMs(), TimeUnit.MILLISECONDS);
+  }
+
+  @Override
+  public byte[] get(byte[] key) {
+    return zkClient.readData(getZkPathForKey(key), true);
+  }
+
+  @Override
+  public void put(byte[] key, byte[] value) {
+    String zkPath = getZkPathForKey(key);
+    createIfNotExists(zkPath);
+    zkClient.writeData(zkPath, value);
+  }
+
+  @Override
+  public void delete(byte[] key) {
+    zkClient.delete(getZkPathForKey(key));
+  }
+
+  @Override
+  public Map<byte[], byte[]> all() {
+    try {
+      List<String> zkSubDirectories = zkClient.getChildren(zkBaseDir);
+      Map<byte[], byte[]> result = new HashMap<>();
+      for (String zkSubDir : zkSubDirectories) {
+        String completeZkPath = String.format("%s/%s", zkBaseDir, zkSubDir);
+        byte[] value = zkClient.readData(completeZkPath, true);
+        result.put(completeZkPath.getBytes("UTF-8"), value);
+      }
+      return result;
+    } catch (Exception e) {
+      LOG.error("Exception occurred on reading the path: {} from zookeeper.", zkBaseDir, e);
+      throw new SamzaException(e);
+    }
+  }
+
+  @Override
+  public void flush() {
+    // No-op for zookeeper implementation.
+  }
+
+  @Override
+  public void close() {
+    zkClient.close();
+  }
+
+  private String getZkPathForKey(byte[] key) {
+    try {
+      return String.format("%s/%s", zkBaseDir, new String(key, "UTF-8"));
+    } catch (UnsupportedEncodingException e) {
+      throw new SamzaException(e);
+    }
+  }
+
+  private void createIfNotExists(String path) {
+    if (!zkClient.exists(path)) {
```

Comment:
Does this operation need to be atomic? Is it OK if multiple callers call createPersistent for the same path? If this is idempotent, do we need to check if it already exists?

Reply:
Zookeeper call `createPersistent` is not idempotent. If the zookeeper path exists, then this API will throw `ZkNodeExistsException`.  

I will make changes, so that this class is thread-safe(So we don't face this exception when multiple callers update the same key at the same time) .

Reply:
Thread safety isn't enough here. This could be called in different processes, so we need to handle ZkNodeExistsException. Can we do the following?
`try { createPersistent(path) } catch (ZkNodeExistsException e) { // ignore }`